beacon/beacon.go

```diff
@@ -135,22 +120,60 @@ func New(
 
 	pd.ctx, pd.cancel = context.WithCancel(pd.ctx)
 	pd.theta = new(big.Float).SetRat(pd.config.Theta)
-	if pd.nonceFetcher == nil {
-		pd.nonceFetcher = defaultFetcher{cdb: cdb}
-	}
 
-	if pd.weakCoin == nil {
-		pd.weakCoin = weakcoin.New(pd.publisher, edSigner.VRFSigner(), vrfVerifier, pd.nonceFetcher, pd,
-			pd.msgTimes,
-			weakcoin.WithLog(pd.logger.WithName("weakCoin")),
-			weakcoin.WithMaxRound(pd.config.RoundsNumber),
-		)
+	if pd.createWeakCoin == nil {
+		pd.createWeakCoin = func(signer *signing.EdSigner) coin {
+			return weakcoin.New(
+				pd.publisher,
+				signer.VRFSigner(),
+				pd.vrfVerifier,
+				pd.nonceFetcher,
+				pd,
+				pd.msgTimes,
+				weakcoin.WithLog(pd.logger.WithName("weakCoin").WithName(signer.NodeID().ShortString())),
+				weakcoin.WithMaxRound(pd.config.RoundsNumber),
+			)
+		}
 	}
-
 	pd.metricsCollector = metrics.NewBeaconMetricsCollector(pd.gatherMetricsData, pd.logger.WithName("metrics"))
 	return pd
 }
 
+func (pd *ProtocolDriver) Register(s *signing.EdSigner) {
+	pd.mu.Lock()
+	defer pd.mu.Unlock()
+	if _, exists := pd.signers[s.NodeID()]; exists {
+		pd.logger.With().Error("signing key already registered", log.ShortStringer("key", s.NodeID()))
+		return
+	}
+	p := participant{
+		signer: s,
+		coin:   pd.createWeakCoin(s),
+	}
+	pd.logger.With().Info("registered signing key", p.Id())
+	pd.signers[s.NodeID()] = p
+}
+
+type participant struct {
+	signer *signing.EdSigner
+	coin   coin
+}
+
+type signerSession struct {
+	participant
+	nonce types.VRFPostIndex
+}
+
+func (s *participant) Id() signerID {
+	return signerID(s.signer.NodeID())
+}
+
+type signerID types.NodeID
+
+func (id signerID) Field() log.Field {
+	return log.ShortStringer("id", types.NodeID(id))
+}
```

Comment:
I'm not sure about introducing a new type here just to change the logging behavour of `NodeID` only for `beacon`. We should consider `(NodeID) Field()` to return a `log.ShortStringer` and have that behaviour everywhere.

Reply:
What's wrong with having a private type here? Btw, it changed the key as well (NodeID logs "node_id", not "id").

Reply:
The problem isn't the private type, it's that `beacon` now logs `NodeID` differently from all other components, which makes it harder to read them imo.